File: src/Client/src/lib/allocationValidator.ts
// ============================================================================
// RAJ Financial – Beneficiary Allocation Validator (client-side)
// ============================================================================
// TypeScript port of Shared/Validators/AllocationValidator.cs.
// Keeps validation logic identical on client and server.
// ============================================================================

import { BeneficiaryAssignmentDto } from "../generated/memorypack/BeneficiaryAssignmentDto";

/* ------------------------------------------------------------------ */
/*  Constants                                                          */
/* ------------------------------------------------------------------ */

export const BeneficiaryType = {
  PRIMARY: "Primary",
  CONTINGENT: "Contingent",
} as const;

export const AllocationErrorCodes = {
  PRIMARY_TOTAL_INVALID: "ALLOCATION_PRIMARY_TOTAL_INVALID",
  CONTINGENT_TOTAL_INVALID: "ALLOCATION_CONTINGENT_TOTAL_INVALID",
  PERCENT_OUT_OF_RANGE: "ALLOCATION_PERCENT_OUT_OF_RANGE",
  DUPLICATE_BENEFICIARY: "ALLOCATION_DUPLICATE_BENEFICIARY",
} as const;

const MIN_PERCENT = 0.01;
const MAX_PERCENT = 100;
const REQUIRED_TOTAL = 100;
const TOLERANCE = 0.0001;

/* ------------------------------------------------------------------ */
/*  Types                                                              */
/* ------------------------------------------------------------------ */

export interface AllocationValidationError {
  errorCode: string;
  message: string;
}

export interface AllocationValidationResult {
  primaryTotal: number;
  contingentTotal: number;
  isPrimaryValid: boolean;
  isContingentValid: boolean;
  errors: AllocationValidationError[];
  isValid: boolean;
}

/* ------------------------------------------------------------------ */
/*  Validator                                                          */
/* ------------------------------------------------------------------ */

function isBeneficiaryType(
  type: string,
  expected: string,
): boolean {
  return type.localeCompare(expected, undefined, { sensitivity: "accent" }) === 0;
}

function findDuplicates(
  assignments: BeneficiaryAssignmentDto[],
  type: string,
  errors: AllocationValidationError[],
): void {
  const seen = new Map<string, string>();
  for (const a of assignments) {
    if (!isBeneficiaryType(a.type, type)) continue;
    if (seen.has(a.beneficiaryId)) {
      errors.push({
        errorCode: AllocationErrorCodes.DUPLICATE_BENEFICIARY,
        message: `'${a.beneficiaryName}' is assigned more than once as a ${type} beneficiary`,
      });
    } else {
      seen.set(a.beneficiaryId, a.beneficiaryName);
    }
  }
}

/**
 * Validates a list of beneficiary assignments against the allocation rules.
 *
 * Rules:
 * - Primary beneficiaries must total exactly 100 %.
 * - Contingent beneficiaries must total exactly 100 % (if any are assigned).
 * - Individual allocations must be between 0.01 % and 100 %.
 * - No duplicate beneficiaries per type.
 */
export function validateAllocations(
  assignments: BeneficiaryAssignmentDto[],
): AllocationValidationResult {
  const errors: AllocationValidationError[] = [];

  // Individual allocation range checks
  for (const a of assignments) {
    if (a.allocationPercent < MIN_PERCENT || a.allocationPercent > MAX_PERCENT) {
      errors.push({
        errorCode: AllocationErrorCodes.PERCENT_OUT_OF_RANGE,
        message: `Allocation for '${a.beneficiaryName}' must be between ${MIN_PERCENT}% and ${MAX_PERCENT}%`,
      });
    }
  }

  // Duplicate checks per type
  findDuplicates(assignments, BeneficiaryType.PRIMARY, errors);
  findDuplicates(assignments, BeneficiaryType.CONTINGENT, errors);

  // Primary total
  const primary = assignments.filter((a) =>
    isBeneficiaryType(a.type, BeneficiaryType.PRIMARY),
  );
  const primaryTotal = primary.reduce((sum, a) => sum + a.allocationPercent, 0);
  const isPrimaryValid = Math.abs(primaryTotal - REQUIRED_TOTAL) < TOLERANCE;

  if (!isPrimaryValid && primary.length > 0) {
    errors.push({
      errorCode: AllocationErrorCodes.PRIMARY_TOTAL_INVALID,
      message: `Primary beneficiary allocations must total exactly 100% (currently ${primaryTotal.toFixed(2)}%)`,
    });
  }

  // Contingent total
  const contingent = assignments.filter((a) =>
    isBeneficiaryType(a.type, BeneficiaryType.CONTINGENT),
  );
  const contingentTotal = contingent.reduce(
    (sum, a) => sum + a.allocationPercent,
    0,
  );
  const isContingentValid =
    contingent.length === 0 ||
    Math.abs(contingentTotal - REQUIRED_TOTAL) < TOLERANCE;

  if (!isContingentValid) {
    errors.push({
      errorCode: AllocationErrorCodes.CONTINGENT_TOTAL_INVALID,
      message: `Contingent beneficiary allocations must total exactly 100% (currently ${contingentTotal.toFixed(2)}%)`,
    });
  }

  return {
    primaryTotal,
    contingentTotal,
    isPrimaryValid,
    isContingentValid,
    errors,
    isValid: errors.length === 0,
  };
}

File: src/Shared/Validators/AllocationValidator.cs
using RajFinancial.Shared.Contracts.Assets;
using RajFinancial.Shared.Entities.Assets;

namespace RajFinancial.Shared.Validators;

/// <summary>
///     Validates beneficiary allocation assignments.
/// </summary>
/// <remarks>
///     <list type="bullet">
///         <item>Primary beneficiaries must total exactly 100%.</item>
///         <item>Contingent beneficiaries must total exactly 100% (if any are assigned).</item>
///         <item>Individual allocations must be between 0.01% and 100%.</item>
///         <item>No duplicate beneficiaries per type.</item>
///     </list>
///     This validator is framework-agnostic and can be used on both server (API) and
///     client (via the TypeScript port in <c>src/Client/src/lib/allocationValidator.ts</c>).
/// </remarks>
public static class AllocationValidator
{
    private const double MIN_PERCENT = 0.01;
    private const double MAX_PERCENT = 100.0;
    private const double REQUIRED_TOTAL = 100.0;
    private const double TOLERANCE = 0.0001;

    /// <summary>
    ///     Validates a list of beneficiary assignments against the allocation rules.
    /// </summary>
    public static AllocationValidationResult Validate(IReadOnlyList<BeneficiaryAssignmentDto> assignments)
    {
        var errors = new List<AllocationValidationError>();

        var primary = assignments
            .Where(a => string.Equals(a.Type, BeneficiaryType.Primary, StringComparison.OrdinalIgnoreCase))
            .ToList();

        var contingent = assignments
            .Where(a => string.Equals(a.Type, BeneficiaryType.Contingent, StringComparison.OrdinalIgnoreCase))
            .ToList();

        // Individual allocation range checks
        foreach (var assignment in assignments)
        {
            if (assignment.AllocationPercent is < MIN_PERCENT or > MAX_PERCENT)
            {
                errors.Add(new AllocationValidationError(
                    AllocationErrorCodes.PercentOutOfRange,
                    $"Allocation for '{assignment.BeneficiaryName}' must be between {MIN_PERCENT}% and {MAX_PERCENT}%"));
            }
        }

        // Duplicate checks per type
        AddDuplicateErrors(primary, BeneficiaryType.Primary, errors);
        AddDuplicateErrors(contingent, BeneficiaryType.Contingent, errors);

        // Primary total must be exactly 100%
        var primaryTotal = primary.Sum(a => a.AllocationPercent);
        var isPrimaryValid = Math.Abs(primaryTotal - REQUIRED_TOTAL) < TOLERANCE;

        if (!isPrimaryValid && primary.Count > 0)
        {
            errors.Add(new AllocationValidationError(
                AllocationErrorCodes.PrimaryTotalInvalid,
                $"Primary beneficiary allocations must total exactly 100% (currently {primaryTotal:F2}%)"));
        }

        // Contingent total must be exactly 100% if any exist
        var contingentTotal = contingent.Sum(a => a.AllocationPercent);
        var isContingentValid = contingent.Count == 0
                                || Math.Abs(contingentTotal - REQUIRED_TOTAL) < TOLERANCE;

        if (!isContingentValid)
        {
            errors.Add(new AllocationValidationError(
                AllocationErrorCodes.ContingentTotalInvalid,
                $"Contingent beneficiary allocations must total exactly 100% (currently {contingentTotal:F2}%)"));
        }

        return new AllocationValidationResult
        {
            PrimaryTotal = primaryTotal,
            ContingentTotal = contingentTotal,
            IsPrimaryValid = isPrimaryValid,
            IsContingentValid = isContingentValid,
            Errors = errors
        };
    }

    private static void AddDuplicateErrors(
        List<BeneficiaryAssignmentDto> assignments,
        string type,
        List<AllocationValidationError> errors)
    {
        var duplicates = assignments
            .GroupBy(a => a.BeneficiaryId)
            .Where(g => g.Count() > 1)
            .Select(g => g.First().BeneficiaryName);

        foreach (var name in duplicates)
        {
            errors.Add(new AllocationValidationError(
                AllocationErrorCodes.DuplicateBeneficiary,
                $"'{name}' is assigned more than once as a {type} beneficiary"));
        }
    }
}
